numpyro/distributions/transforms.py

```diff
@@ -129,6 +129,91 @@ def log_abs_det_jacobian(self, x, y):
+    r = r + np.identity(r.shape[-1])
+
+    # apply stick-breaking on the squared values;
+    # we omit the step of computing s = z * z_cumprod by using the fact:
+    #     y = sign(r) * s = sign(r) * sqrt(z * z_cumprod) = r * sqrt(z_cumprod)
+    z = r ** 2
+    z1m_cumprod = cumprod(1 - z)
+
+    # to workaround the issue: NaN propagated through backward pass even when not accessed
+    # at https://github.com/pytorch/pytorch/issues/15506 (which also happens in JAX),
+    # here we only take sqrt at tril part
+    z1m_cumprod_tril_sqrt = np.sqrt(matrix_to_tril_vec(z1m_cumprod, diagonal=-1))
+    z1m_cumprod_sqrt = vec_to_tril_matrix(z1m_cumprod_tril_sqrt, diagonal=-1)
+
+    pad_width = [(0, 0)] * z.ndim
+    pad_width[-1] = (1, 0)
+    z1m_cumprod_sqrt_shifted = np.pad(z1m_cumprod_sqrt[..., :-1], pad_width,
+                                      mode="constant", constant_values=1.)
+    y = r * z1m_cumprod_sqrt_shifted
+    return y
+
+
+class CorrCholeskyTransform(Transform):
+    r"""
+    Transforms a uncontrained real vector :math:`x` with length :math:`D*(D-1)/2` into the
+    Cholesky factor of a D-dimension correlation matrix. This Cholesky factor is a lower
+    triangular matrix with positive diagonals and unit Euclidean norm for each row.
+    The transform is processed as follows:
+    1. First we convert :math:`x` into a lower triangular matrix with the following order:
+    .. math::
+        \begin{bmatrix}
+            1   & 0 & 0 & 0 \\
+            x_0 & 1 & 0 & 0 \\
+            x_1 & x_2 & 1 & 0 \\
+            x_3 & x_4 & x_5 & 1
+        \end{bmatrix}
+    2. For each row :math:`X_i` of the lower triangular part, we apply a *signed* version of
+    class :class:`StickBreakingTransform` to transform :math:`X_i` into a
+    unit Euclidean length vector using the following steps:
+        a. Scales into the interval :math:`(-1, 1)` domain: :math:`r_i = \tanh(X_i)`.
+        b. Transforms into an unsigned domain: :math:`z_i = r_i^2`.
+        c. Applies :math:`s_i = StickBreakingTransform(z_i)`.
+        d. Transforms back into signed domain: :math:`y_i = (sign(r_i), 1) * \sqrt{s_i}`.
+    """
+    codomain = constraints.corr_cholesky
+    event_dim = 1
+
+    def __call__(self, x):
+        # we interchange step 1 and step 2.a for a better performance
+        eps = np.finfo(x.dtype).eps
+        t = np.clip(np.tanh(x), a_min=(-1 + eps), a_max=(1 - eps))
+        return _signed_stick_breaking_tril(t)
+
+    def inv(self, y):
+        # inverse stick-breaking
+        z1m_cumprod = 1 - cumsum(y * y)
+        pad_width = [(0, 0)] * y.ndim
+        pad_width[-1] = (1, 0)
+        z1m_cumprod_shifted = np.pad(z1m_cumprod[..., :-1], pad_width,
+                                     mode="constant", constant_values=1.)
+        t = matrix_to_tril_vec(y, diagonal=-1) / np.sqrt(
+            matrix_to_tril_vec(z1m_cumprod_shifted, diagonal=-1))
+        # inverse of tanh
+        x = np.log((1 + t) / (1 - t)) / 2
```

Comment:
Can we test how numerically stable this is, e.g. by applying the forward and inverse transforms? The LKJ implementation in Pyro brought up issues with numerical stability, but I do not recall what they were.

Reply:
We did this test for all transforms at [this line](https://github.com/pyro-ppl/numpyro/pull/126/files#diff-51c2d9ee9d114f18441e3579ec31fdbeR466).

Reply:
For some reason, I can't reply to the `slogdet` comment. In GPU, `onp.linalg.slogdet` returns a numpy array.